<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.269">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Jimmi">

<title>Class 18: Mutational Signatures in Human Cancer</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="class18_files/libs/clipboard/clipboard.min.js"></script>
<script src="class18_files/libs/quarto-html/quarto.js"></script>
<script src="class18_files/libs/quarto-html/popper.min.js"></script>
<script src="class18_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="class18_files/libs/quarto-html/anchor.min.js"></script>
<link href="class18_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="class18_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="class18_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="class18_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="class18_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Class 18: Mutational Signatures in Human Cancer</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Jimmi </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<blockquote class="blockquote">
<p>Q. How many cancer samples are included in the dataset?</p>
</blockquote>
<p>448 cancer samples were included.</p>
<blockquote class="blockquote">
<p>Q. Which is the most mutated gene?</p>
</blockquote>
<p>The TTN gene was the most mutated gene with 80% frequency in melonoma patients.</p>
<blockquote class="blockquote">
<p>Q. Which is the most common treatment undergone by patients?</p>
</blockquote>
<p>The most common treatment undergone is Radiation 1.</p>
<section id="generating-mutational-matrices-and-visualizing-mutational-profiles" class="level2">
<h2 class="anchored" data-anchor-id="generating-mutational-matrices-and-visualizing-mutational-profiles">Generating mutational matrices and visualizing mutational profiles</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Install required packages</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">"BiocManager"</span>)){</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: BiocManager</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in file(con, "r"): URL 'https://bioconductor.org/config.yaml': status
was 'Couldn't resolve host name'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in file(con, "r"): URL 'http://bioconductor.org/config.yaml': status
was 'Couldn't resolve host name'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Bioconductor version cannot be validated; no internet connection?  See
  #troubleshooting section in vignette</code></pre>
</div>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">"maftools"</span>)){</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>    BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"maftools"</span>)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: maftools</code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">"BSgenome.Hsapiens.UCSC.hg19"</span>)){         <span class="co"># reference genome needed to</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"BSgenome.Hsapiens.UCSC.hg19"</span>)   <span class="co"># generate mutational matrices</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: BSgenome.Hsapiens.UCSC.hg19</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: BSgenome</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: BiocGenerics</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    IQR, mad, sd, var, xtabs</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    anyDuplicated, aperm, append, as.data.frame, basename, cbind,
    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,
    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,
    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,
    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,
    table, tapply, union, unique, unsplit, which.max, which.min</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: S4Vectors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: stats4</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'S4Vectors'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    expand.grid, I, unname</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: IRanges</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'IRanges'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:grDevices':

    windows</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: GenomeInfoDb</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: GenomicRanges</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: Biostrings</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: XVector</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'Biostrings'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:base':

    strsplit</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: rtracklayer</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read maf file</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(maftools)</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>coad <span class="ot">=</span> <span class="fu">read.maf</span>(<span class="st">'skcm_tcga_pan_can_atlas_2018/data_mutations.txt'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>-Reading
-Validating
--Removed 27563 duplicated variants
-Silent variants: 209854 
-Summarizing
--Possible FLAGS among top ten genes:
  TTN
  MUC16
-Processing clinical data
--Missing clinical data
-Finished in 36.9s elapsed (21.2s cpu) </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate mutational matrix (SBS96 context)</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>mm_coad <span class="ot">=</span> <span class="fu">trinucleotideMatrix</span>(<span class="at">maf =</span> coad, <span class="at">prefix =</span> <span class="st">'chr'</span>, <span class="at">add =</span> <span class="cn">TRUE</span>,</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>                              <span class="at">ref_genome =</span> <span class="st">"BSgenome.Hsapiens.UCSC.hg19"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>-Extracting 5' and 3' adjacent bases
-Extracting +/- 20bp around mutated bases for background C&gt;T estimation
-Estimating APOBEC enrichment scores
--Performing one-way Fisher's test for APOBEC enrichment
---APOBEC related mutations are enriched in  1.818 % of samples (APOBEC enrichment score &gt; 2 ;  8  of  440  samples)
-Creating mutation matrix
--matrix of dimension 440x96</code></pre>
</div>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>mm_coad <span class="ot">=</span> <span class="fu">t</span>(mm_coad<span class="sc">$</span>nmf_matrix)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For the visualization of SBS96 mutational profiles, we will make use of the MutationalPatterns R package. This library is commonly used for all kinds of mutational signature analysis, and we will also use it for the subsequent assignment analysis.</p>
<p>To generate mutational profiles, we will use the previously generated mutational matrices as input. As the cohorts are pretty large, we will focus on exploring the mutational profiles of 4 random samples, the 4 samples harboring the highest number of mutations, the 4 samples with the lowest number of mutations, and the average mutational profile of all samples of the cohort.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>mm_coad <span class="ot">=</span> <span class="fu">read.delim</span>(<span class="st">'skin.txt'</span>)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Install MutationalPatterns package</span></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">"MutationalPatterns"</span>)){</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"MutationalPatterns"</span>)</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: MutationalPatterns</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: NMF</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: registry</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: rngtools</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: cluster</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>NMF - BioConductor layer [OK] | Shared memory capabilities [NO: windows] | Cores 7/8</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'NMF'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:S4Vectors':

    nrun</code></pre>
</div>
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate mutational profiles (4 random samples)</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(MutationalPatterns)</span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">11111</span>) <span class="co"># fixing the seed for random number generation</span></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>samples_to_plot <span class="ot">=</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(mm_coad),<span class="dv">4</span>) <span class="co"># selecting 4 random samples</span></span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_96_profile</span>(mm_coad[,samples_to_plot], <span class="at">condensed =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="class18_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate mutational profiles (top 4 mutated samples and top 4 less mutated)</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>mutations_in_samples <span class="ot">=</span> <span class="fu">colSums</span>(mm_coad)</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>mutations_in_samples <span class="ot">=</span> <span class="fu">sort</span>(mutations_in_samples, <span class="at">decreasing =</span> T)</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>samples_to_plot <span class="ot">=</span> <span class="fu">names</span>(mutations_in_samples)[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>]</span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_96_profile</span>(mm_coad[,samples_to_plot], <span class="at">condensed =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="class18_files/figure-html/unnamed-chunk-4-2.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>mutations_in_samples <span class="ot">=</span> <span class="fu">sort</span>(mutations_in_samples, <span class="at">decreasing =</span> F)</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>samples_to_plot <span class="ot">=</span> <span class="fu">names</span>(mutations_in_samples)[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>]</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_96_profile</span>(mm_coad[,samples_to_plot], <span class="at">condensed =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="class18_files/figure-html/unnamed-chunk-4-3.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate average mutational profiles</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>relative_mutational_profile <span class="ot">=</span> <span class="fu">apply</span>(mm_coad, <span class="dv">2</span>, prop.table) <span class="co"># obtained relative</span></span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>                                                            <span class="co"># mutational matrix</span></span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>average_mutational_profile <span class="ot">=</span> <span class="fu">rowMeans</span>(relative_mutational_profile)</span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a>average_mutational_profile <span class="ot">=</span> <span class="fu">data.frame</span>(average_mutational_profile)</span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_96_profile</span>(average_mutational_profile, <span class="at">condensed =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="class18_files/figure-html/unnamed-chunk-4-4.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="assigning-reference-mutational-signatures" class="level2">
<h2 class="anchored" data-anchor-id="assigning-reference-mutational-signatures">Assigning reference mutational signatures</h2>
<p>Leveraging the COSMIC mutational signatures, we will perform a mutational signature assignment analysis to quantify the number of mutations contributed by each signature to a given cancer sample and, therefore, decipher which mutational processes have been active in each individual tumor.</p>
<p>We will perform two different assignment analyses using the <code>MutationalPatterns</code> package</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Mutational signature assignment</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>cosmic_signatures <span class="ot">=</span> <span class="fu">get_known_signatures</span>(<span class="at">source =</span> <span class="st">'COSMIC_v3.2'</span>)</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>fit_res <span class="ot">=</span> <span class="fu">fit_to_signatures</span>(mm_coad, cosmic_signatures)</span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Top contributing signatures</span></span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>contributions <span class="ot">=</span> fit_res<span class="sc">$</span>contribution</span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a>top_contributing_signatures_abs <span class="ot">=</span> <span class="fu">rowMeans</span>(contributions)</span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a>top_contributing_signatures_abs <span class="ot">=</span> <span class="fu">sort</span>(top_contributing_signatures_abs,</span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a>                                       <span class="at">decreasing =</span> T)[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>]</span>
<span id="cb47-11"><a href="#cb47-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-12"><a href="#cb47-12" aria-hidden="true" tabindex="-1"></a><span class="do">## Top 4 contributing signatures (absolute values)</span></span>
<span id="cb47-13"><a href="#cb47-13" aria-hidden="true" tabindex="-1"></a>top_contributing_signatures_abs</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    SBS7a     SBS7b     SBS38      SBS4 
366.97614 340.91011 204.44450  99.49106 </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>relative_contributions <span class="ot">=</span> <span class="fu">apply</span>(contributions,<span class="dv">2</span>,prop.table)</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>top_contributing_signatures_rel <span class="ot">=</span> <span class="fu">rowMeans</span>(relative_contributions)</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>top_contributing_signatures_rel <span class="ot">=</span> <span class="fu">sort</span>(top_contributing_signatures_rel,</span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>                                       <span class="at">decreasing =</span> T)[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>]</span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a><span class="do">## Top 4 contributing signatures (relative values)</span></span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a>top_contributing_signatures_rel</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     SBS7b      SBS7a      SBS38       SBS4 
0.26336351 0.26019455 0.10885595 0.07240978 </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Mutational signature assignment strict</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>fit_res_strict <span class="ot">=</span> <span class="fu">fit_to_signatures_strict</span>(mm_coad, cosmic_signatures)</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>fit_res_strict <span class="ot">=</span> fit_res_strict<span class="sc">$</span>fit_res</span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>contributions_strict <span class="ot">=</span> fit_res_strict<span class="sc">$</span>contribution</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="visualizing-mutational-signature-assignment-results" class="level2">
<h2 class="anchored" data-anchor-id="visualizing-mutational-signature-assignment-results">Visualizing mutational signature assignment results</h2>
<p>To visualize the mutational signature assignment results, we will use the default visualizations available in the <code>MutationalPatterns</code> package.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualization of signature assignment results (fit_to_signatures)</span></span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">11111</span>)</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>samples_to_plot <span class="ot">=</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(mm_coad),<span class="dv">4</span>)</span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_contribution</span>(contributions[,samples_to_plot], <span class="at">mode =</span> <span class="st">"absolute"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="class18_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_contribution</span>(contributions[,samples_to_plot], <span class="at">mode =</span> <span class="st">"relative"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="class18_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_contribution_heatmap</span>(contributions, <span class="at">cluster_samples =</span> F)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="class18_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualization of signature assignment results (strict)</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_contribution</span>(contributions_strict[,samples_to_plot], <span class="at">mode =</span> <span class="st">"absolute"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="class18_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_contribution</span>(contributions_strict[,samples_to_plot], <span class="at">mode =</span> <span class="st">"relative"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="class18_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_contribution_heatmap</span>(contributions_strict, <span class="at">cluster_samples =</span> F)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="class18_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>To check the cosine similarity of the reconstruction for some specific samples, we can use the following visualization from the <code>MutationalPatterns</code> R package.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Cosine similarity reconstruction vs. original mutational profile (fit_to_signatures)</span></span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">11111</span>)</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>samples_to_plot <span class="ot">=</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(mm_coad),<span class="dv">4</span>)</span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_original_vs_reconstructed</span>(mm_coad[,samples_to_plot],</span>
<span id="cb58-6"><a href="#cb58-6" aria-hidden="true" tabindex="-1"></a>                               fit_res<span class="sc">$</span>reconstructed[,samples_to_plot], </span>
<span id="cb58-7"><a href="#cb58-7" aria-hidden="true" tabindex="-1"></a>                               <span class="at">y_intercept =</span> <span class="fl">0.90</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="class18_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Cosine similarity reconstruction vs. original mutational profile (strict)</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_original_vs_reconstructed</span>(mm_coad[,samples_to_plot],</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>                               fit_res_strict<span class="sc">$</span>reconstructed[,samples_to_plot], </span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a>                               <span class="at">y_intercept =</span> <span class="fl">0.90</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="class18_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<blockquote class="blockquote">
<p>Q. Which is the etiology of the top absolute contributing signature for liver cancer?</p>
</blockquote>
<p>Aristolochic acid exposure</p>
<blockquote class="blockquote">
<p>Q. Which is the most prominent mutational context for the top contributing signature in skin cancer?</p>
</blockquote>
<p>C&gt;T</p>
<blockquote class="blockquote">
<p>Q. The etiology of the top contributing signature for lung cancer corresponds to an endogenous cellular mechanism.</p>
</blockquote>
<p>False, Tobacco smoking is the etiology.</p>
<blockquote class="blockquote">
<p>Q. SBS4 is one of the most common signatures found in lung cancer and is associated with tobacco smoking.</p>
</blockquote>
<p>True, it is the most common signature.</p>
<blockquote class="blockquote">
<p>Q. SBS7d is one of the most common signatures in skin cancer and is associated with UV light exposure and high numbers of C&gt;T mutations.</p>
</blockquote>
<p>False, SBS7A is common signature.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.2.2 (2022-10-31 ucrt)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 22621)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.utf8 
[2] LC_CTYPE=English_United States.utf8   
[3] LC_MONETARY=English_United States.utf8
[4] LC_NUMERIC=C                          
[5] LC_TIME=English_United States.utf8    

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] MutationalPatterns_3.8.1          NMF_0.25                         
 [3] Biobase_2.58.0                    cluster_2.1.4                    
 [5] rngtools_1.5.2                    registry_0.5-1                   
 [7] BSgenome.Hsapiens.UCSC.hg19_1.4.3 BSgenome_1.66.3                  
 [9] rtracklayer_1.58.0                Biostrings_2.66.0                
[11] XVector_0.38.0                    GenomicRanges_1.50.2             
[13] GenomeInfoDb_1.34.9               IRanges_2.32.0                   
[15] S4Vectors_0.36.2                  BiocGenerics_0.44.0              
[17] maftools_2.14.0                   BiocManager_1.30.20              

loaded via a namespace (and not attached):
 [1] MatrixGenerics_1.10.0       tidyr_1.3.0                
 [3] jsonlite_1.8.4              splines_4.2.2              
 [5] foreach_1.5.2               GenomeInfoDbData_1.2.9     
 [7] Rsamtools_2.14.0            yaml_2.3.7                 
 [9] pillar_1.8.1                lattice_0.20-45            
[11] glue_1.6.2                  digest_0.6.31              
[13] RColorBrewer_1.1-3          colorspace_2.1-0           
[15] htmltools_0.5.4             Matrix_1.5-1               
[17] plyr_1.8.8                  XML_3.99-0.13              
[19] pkgconfig_2.0.3             zlibbioc_1.44.0            
[21] purrr_1.0.1                 scales_1.2.1               
[23] ggalluvial_0.12.5           pracma_2.4.2               
[25] BiocParallel_1.32.5         tibble_3.1.8               
[27] farver_2.1.1                generics_0.1.3             
[29] ggplot2_3.4.1               withr_2.5.0                
[31] SummarizedExperiment_1.28.0 cli_3.6.0                  
[33] survival_3.4-0              magrittr_2.0.3             
[35] crayon_1.5.2                evaluate_0.20              
[37] fansi_1.0.4                 doParallel_1.0.17          
[39] tools_4.2.2                 data.table_1.14.8          
[41] BiocIO_1.8.0                lifecycle_1.0.3            
[43] matrixStats_0.63.0          gridBase_0.4-7             
[45] stringr_1.5.0               munsell_0.5.0              
[47] DelayedArray_0.23.2         compiler_4.2.2             
[49] rlang_1.0.6                 grid_4.2.2                 
[51] RCurl_1.98-1.10             iterators_1.0.14           
[53] rstudioapi_0.14             rjson_0.2.21               
[55] htmlwidgets_1.6.1           labeling_0.4.2             
[57] bitops_1.0-7                rmarkdown_2.20             
[59] DNAcopy_1.72.3              restfulr_0.0.15            
[61] gtable_0.3.1                codetools_0.2-18           
[63] DBI_1.1.3                   reshape2_1.4.4             
[65] R6_2.5.1                    GenomicAlignments_1.34.0   
[67] knitr_1.42                  dplyr_1.1.0                
[69] fastmap_1.1.1               utf8_1.2.3                 
[71] stringi_1.7.12              parallel_4.2.2             
[73] Rcpp_1.0.10                 vctrs_0.5.2                
[75] tidyselect_1.2.0            xfun_0.37                  </code></pre>
</div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>